File: bindings/zig/ove/src/workqueue.zig
// Copyright (C) 2026 Kamil Lulko <kamil.lulko@gmail.com>
//
// SPDX-License-Identifier: GPL-3.0-or-later
//
// This file is part of oveRTOS.

const std = @import("std");
const c = @import("c.zig").raw;
const err = @import("error.zig");
const Error = err.Error;
const thread_mod = @import("thread.zig");
const Duration = @import("time.zig").Duration;

/// Deferred work queue backed by a dedicated RTOS thread.
///
/// `stack_size` is a comptime type parameter so the worker stack can be
/// laid out inline in the allocator-managed backing struct.  Both heap
/// and zero-heap modes use the substrate's `_init` path against a stack
/// + storage block drawn from the user's `std.mem.Allocator`.
///
/// ```zig
/// var wq = try ove.Workqueue(2048).create(allocator, "bench_wq", .normal);
/// defer wq.deinit();
/// ```
pub fn Workqueue(comptime stack_size: usize) type {
    return struct {
        const Self = @This();

        // Zephyr MPU requires power-of-2 alignment + 128-byte guard pad
        // on every kernel thread stack — reuse the thread module's
        // helpers so the embedded stack matches what `ove_workqueue_init`
        // expects.
        const Backing = struct {
            stack: [thread_mod.stackTotal(stack_size)]u8 align(thread_mod.stackAlign(stack_size)),
            storage: c.ove_workqueue_storage_t,
        };

        allocator: std.mem.Allocator,
        handle: c.ove_workqueue_t,
        backing: *Backing,

        /// Allocate the worker stack + substrate-storage from
        /// `allocator` and spawn a dedicated thread running
        /// `ove_workqueue_init` at `priority`.
        pub fn create(
            allocator: std.mem.Allocator,
            name: [*:0]const u8,
            priority: thread_mod.Priority,
        ) Error!Self {
            const backing = try allocator.create(Backing);
            errdefer allocator.destroy(backing);
            backing.stack = std.mem.zeroes(@TypeOf(backing.stack));
            backing.storage = std.mem.zeroes(c.ove_workqueue_storage_t);
            var h: c.ove_workqueue_t = null;
            try err.fromCode(c.ove_workqueue_init(
                &h,
                &backing.storage,
                name,
                @intFromEnum(priority),
                stack_size,
                &backing.stack,
            ));
            return .{ .allocator = allocator, .handle = h, .backing = backing };
        }

        pub fn deinit(self: Self) void {
            if (self.handle != null) c.ove_workqueue_deinit(self.handle);
            self.allocator.destroy(self.backing);
        }

        /// Enqueue `work` for execution by the workqueue thread.
        pub fn submit(self: Self, work: *Work) Error!void {
            try err.fromCode(c.ove_work_submit(self.handle, work.handle));
        }

        /// Submit `work` with a delay.  Substrate takes the delay as
        /// milliseconds; saturating-cast from the `Duration` ns count.
        pub fn submitDelayedFor(self: Self, work: *Work, d: Duration) Error!void {
            const ms: u32 = if (d.ns / std.time.ns_per_ms > std.math.maxInt(u32))
                std.math.maxInt(u32)
            else
                @intCast(d.ns / std.time.ns_per_ms);
            try err.fromCode(c.ove_work_submit_delayed(self.handle, work.handle, ms));
        }
    };
}

/// A single deferred work item that wraps a Zig callback.
///
/// ```zig
/// var work = try ove.Work.create(allocator, myHandler);
/// defer work.deinit();
/// try wq.submit(&work);
/// ```
pub const Work = struct {
    allocator: std.mem.Allocator,
    handle: c.ove_work_t,
    storage: *c.ove_work_storage_t,

    /// Create a deferred work item bound to a plain Zig callback.
    pub fn create(allocator: std.mem.Allocator, comptime handler: fn () void) Error!Work {
        const Tramp = struct {
            fn invoke(_: c.ove_work_t) callconv(.c) void {
                handler();
            }
        };
        const storage = try allocator.create(c.ove_work_storage_t);
        errdefer allocator.destroy(storage);
        storage.* = std.mem.zeroes(c.ove_work_storage_t);
        var h: c.ove_work_t = null;
        try err.fromCode(c.ove_work_init_static(&h, storage, &Tramp.invoke));
        return .{ .allocator = allocator, .handle = h, .storage = storage };
    }

    /// Create with a typed context pointer.  `ctx` must outlive the work.
    pub fn createWithContext(
        allocator: std.mem.Allocator,
        comptime Context: type,
        ctx: *Context,
        comptime handler: fn (*Context) void,
    ) Error!Work {
        const Captured = struct {
            var ctx_ptr: ?*Context = null;
            fn invoke(_: c.ove_work_t) callconv(.c) void {
                if (ctx_ptr) |p| handler(p);
            }
        };
        Captured.ctx_ptr = ctx;
        const storage = try allocator.create(c.ove_work_storage_t);
        errdefer allocator.destroy(storage);
        storage.* = std.mem.zeroes(c.ove_work_storage_t);
        var h: c.ove_work_t = null;
        try err.fromCode(c.ove_work_init_static(&h, storage, &Captured.invoke));
        return .{ .allocator = allocator, .handle = h, .storage = storage };
    }

    pub fn deinit(self: Work) void {
        // Static work items have no substrate-level deinit; we just
        // release our storage.
        self.allocator.destroy(self.storage);
    }

    /// Cancel a pending submission.  No-op if the work has already
    /// run or is currently executing.
    pub fn cancel(self: Work) Error!void {
        try err.fromCode(c.ove_work_cancel(self.handle));
    }
};

File: bindings/rust/ove/src/workqueue.rs
// Copyright (C) 2026 Kamil Lulko <kamil.lulko@gmail.com>
//
// SPDX-License-Identifier: GPL-3.0-or-later
//
// This file is part of oveRTOS.

//! Work queue and deferred work items for oveRTOS.
//!
//! A [`Workqueue`] runs a dedicated thread that processes [`Work`] items in order.
//! Work items can be submitted immediately or after a delay, and can be cancelled
//! before they execute.

use crate::bindings;
use crate::error::{Error, Result};

// SAFETY (module-wide contract for the `unsafe { bindings::ove_*(...) }` FFI
// calls below): any handle passed to the C API is non-null and refers to a
// live RTOS object — wrapper constructors establish validity via
// `Error::from_code`, and `Drop` (or an explicit `deinit`) is the only place
// a handle is released. Pointer and slice arguments reference caller-owned
// memory valid for the duration of the call; the C side copies whatever it
// retains and does not alias them past return (verified against the
// signatures in `include/ove/*.h`). Blocks that deviate — `transmute`, raw
// pointer casts from user data, slice reconstruction via `from_raw_parts`,
// or storing a callback across the FFI boundary — carry their own
// `// SAFETY:` comment.

/// A workqueue that executes deferred work items on a dedicated thread.
pub struct Workqueue {
    handle: bindings::ove_workqueue_t,
}

impl Workqueue {
    /// Create a workqueue via heap allocation (only in heap mode).
    ///
    /// `name` must be a `&'static CStr`; the easiest source is a
    /// `c"..."` literal (Rust 1.77+).  The compiler enforces null
    /// termination — `&[u8]` without `\0` was the previous shape and
    /// was undefined-behaviour-prone.
    #[cfg(not(zero_heap))]
    pub fn new(
        name: &'static core::ffi::CStr,
        priority: crate::Priority,
        stack_size: usize,
    ) -> Result<Self> {
        let mut handle: bindings::ove_workqueue_t = core::ptr::null_mut();
        let rc = unsafe {
            bindings::ove_workqueue_create(
                &mut handle,
                name.as_ptr(),
                priority as bindings::ove_prio_t,
                stack_size,
            )
        };
        Error::from_code(rc)?;
        Ok(Self { handle })
    }

    /// Create from caller-provided static storage and stack.
    ///
    /// # Safety
    /// - `storage` must outlive the `Workqueue` and not be shared.
    /// - `stack` must point to at least `stack_size` bytes and outlive
    ///   the `Workqueue`.
    #[cfg(zero_heap)]
    pub unsafe fn from_static(
        storage: *mut bindings::ove_workqueue_storage_t,
        name: &'static core::ffi::CStr,
        priority: crate::Priority,
        stack_size: usize,
        stack: *mut core::ffi::c_void,
    ) -> Result<Self> {
        let mut handle: bindings::ove_workqueue_t = core::ptr::null_mut();
        let rc = unsafe {
            bindings::ove_workqueue_init(
                &mut handle,
                storage,
                name.as_ptr(),
                priority as bindings::ove_prio_t,
                stack_size,
                stack,
            )
        };
        Error::from_code(rc)?;
        Ok(Self { handle })
    }

    /// Return the underlying `ove_workqueue_t` handle for advanced
    /// FFI interop.  Normal app code should pass the [`Workqueue`]
    /// itself to [`Work::submit`] / [`Work::submit_delayed`].
    pub fn handle(&self) -> bindings::ove_workqueue_t {
        self.handle
    }
}

crate::ove_handle_impl!(Workqueue, ove_workqueue_destroy, ove_workqueue_deinit);

/// A work item that can be submitted to a [`Workqueue`].
pub struct Work {
    handle: bindings::ove_work_t,
}

impl Work {
    /// Create a work item via heap allocation (only in heap mode).
    ///
    /// The handler receives the raw `ove_work_t` handle (the C API does
    /// not provide a user_data parameter for work handlers).
    ///
    /// # Errors
    /// Returns [`Error::NoMemory`] if heap allocation fails.
    #[cfg(not(zero_heap))]
    pub fn new(handler: bindings::ove_work_fn) -> Result<Self> {
        let mut handle: bindings::ove_work_t = core::ptr::null_mut();
        let rc = unsafe { bindings::ove_work_init(&mut handle, handler) };
        Error::from_code(rc)?;
        Ok(Self { handle })
    }

    /// Create from caller-provided static storage.
    ///
    /// # Safety
    /// Caller must ensure `storage` outlives the `Work` item.
    #[cfg(zero_heap)]
    pub unsafe fn from_static(
        storage: *mut bindings::ove_work_storage_t,
        handler: bindings::ove_work_fn,
    ) -> Result<Self> {
        let mut handle: bindings::ove_work_t = core::ptr::null_mut();
        let rc = unsafe { bindings::ove_work_init_static(&mut handle, storage, handler) };
        Error::from_code(rc)?;
        Ok(Self { handle })
    }

    /// Submit this work item to `wq` for immediate execution.
    ///
    /// # Errors
    /// Returns an error if the workqueue is shutting down or the item is already pending.
    pub fn submit(&self, wq: &Workqueue) -> Result<()> {
        let rc = unsafe { bindings::ove_work_submit(wq.handle(), self.handle) };
        Error::from_code(rc)
    }

    /// Submit this work item to `wq` for execution after `delay_ms` milliseconds.
    ///
    /// # Errors
    /// Returns an error if the workqueue is shutting down or the item is already pending.
    pub fn submit_delayed(&self, wq: &Workqueue, delay_ms: u32) -> Result<()> {
        let rc = unsafe { bindings::ove_work_submit_delayed(wq.handle(), self.handle, delay_ms) };
        Error::from_code(rc)
    }

    /// Cancel this work item if it is pending or delayed.
    ///
    /// Has no effect if the item is not currently queued.
    ///
    /// # Errors
    /// Returns an error if the underlying RTOS call fails.
    pub fn cancel(&self) -> Result<()> {
        let rc = unsafe { bindings::ove_work_cancel(self.handle) };
        Error::from_code(rc)
    }
}

impl Drop for Work {
    fn drop(&mut self) {
        if self.handle.is_null() {
            return;
        }
        #[cfg(not(zero_heap))]
        unsafe {
            bindings::ove_work_free(self.handle);
        }
        #[cfg(zero_heap)]
        unsafe {
            bindings::ove_work_deinit(self.handle);
        }

        self.handle = core::ptr::null_mut();
    }
}

// SAFETY: `Work` wraps `ove_work_t`.  Submission and cancellation are
// substrate-synchronised; the handler trampoline runs on the workqueue
// thread, never the constructing thread, so cross-thread transfer is
// the expected access pattern.
unsafe impl Send for Work {}
unsafe impl Sync for Work {}
